ts/nni_manager/core/tuner_command_channel/websocket_channel.ts

```diff
@@ -0,0 +1,146 @@
+ *   3. Refactor message format.
+ **/
+
+import assert from 'assert/strict';
+import { EventEmitter } from 'events';
+
+import { Deferred } from 'ts-deferred';
+import type WebSocket from 'ws';
+
+import { Logger, getLogger } from 'common/log';
+
+const logger: Logger = getLogger('tuner_command_channel.WebSocketChannel');
+
+export interface WebSocketChannel {
+    init(): Promise<void>;
+    shutdown(): Promise<void>;
+    sendCommand(command: string): void;  // maybe this should return Promise<void>
+    onCommand(callback: (command: string) => void): void;
+    onError(callback: (error: Error) => void): void;
+}
+
+/**
+ *  Get the singleton tuner command channel.
+ *  Remember to invoke ``await channel.init()`` before doing anything else.
+ **/
+export function getWebSocketChannel(): WebSocketChannel {
+    return channelSingleton;
+}
+
+/**
+ *  The callback to serve WebSocket connection request. Used by REST server module.
+ *  It should only be invoked once, or an error will be raised.
+ *
+ *  Type hint of express-ws is somewhat problematic. Don't want to waste time on it so use `any`.
+ **/
+export function serveWebSocket(ws: WebSocket): void {
+    channelSingleton.setWebSocket(ws);
+}
+
+class WebSocketChannelImpl {
+    private deferredInit: Deferred<void> = new Deferred<void>();
+    private emitter: EventEmitter = new EventEmitter();
+    private heartbeatTimer!: NodeJS.Timer;
+    private serving: boolean = false;
+    private waitingPong: boolean = false;
+    private ws!: WebSocket;
+
+    public setWebSocket(ws: WebSocket): void {
+        if (this.ws !== undefined) {
+            logger.error('A second client is trying to connect');
+            ws.close(4030, 'Already serving a tuner.');
+            return;
+        }
+
+        logger.debug('Connected.');
+        this.serving = true;
+
+        this.ws = ws;
+        ws.on('close', () => { this.handleError(new Error('tuner_command_channel: Tuner closed connection')); });
+        ws.on('error', this.handleError.bind(this));
+        ws.on('message', this.receive.bind(this));
+        ws.on('pong', () => { this.waitingPong = false; });
+
+        this.heartbeatTimer = setInterval(this.heartbeat.bind(this), heartbeatInterval);
+        this.deferredInit.resolve();
+    }
+
+    public init(): Promise<void> {
+        logger.debug(this.ws === undefined ? 'Waiting connection...' : 'Initialized.');
+        return this.deferredInit.promise;
+    }
+
+    public async shutdown(): Promise<void> {
+        if (this.ws === undefined) {
+            return;
+        }
+        clearInterval(this.heartbeatTimer);
+        this.serving = false;
+        this.emitter.removeAllListeners();
+    }
+
+    public sendCommand(command: string): void {
+        assert.ok(this.ws !== undefined);
+
+        logger.debug('Sending', command);
+        this.ws.send(command);
+
+        if (this.ws.bufferedAmount > command.length + 1000) {
```

Comment:
Is this 1000 a magic number?

Reply:
Yes. It's only for a warning message. Any value above zero is acceptable.